rust/chains/abacus-ethereum/src/trait_builder.rs

```diff
@@ -0,0 +1,59 @@
+use std::sync::Arc;
+
+use ethers::prelude::*;
+
+use abacus_core::{ContractLocator, Signers};
+
+use crate::Connection;
+
+async fn make_signing_provider<M: Middleware>(
+    provider: M,
+    signer: Signers,
+) -> Result<SignerMiddleware<NonceManagerMiddleware<M>, Signers>, M::Error> {
+    let provider_chain_id = provider.get_chainid().await?;
+    let signer = ethers::signers::Signer::with_chain_id(signer, provider_chain_id.as_u64());
+
+    let address = ethers::prelude::Signer::address(&signer);
+    let provider = NonceManagerMiddleware::new(provider, address);
+
+    let signing_provider = SignerMiddleware::new(provider, signer);
+    Ok(signing_provider)
+}
+
+pub(crate) trait MakeableWithProvider {
+    type Output;
+
+    fn make<M: Middleware + 'static>(self, provider: M, locator: &ContractLocator) -> Self::Output;
+}
+
+pub(super) async fn build_trait<M: MakeableWithProvider>(
```

Comment:
Should this function just go in the `MakeableWithProvider` trait? And then the helper functions in rust/chains/abacus-ethereum/src/lib.rs could be made superflous?

Reply:
That would probably be smart. I don't see any issue with that at the moment.